core/templates/pages/exploration-editor-page/editor-tab/services/responses.service.ts

```diff
@@ -120,6 +119,15 @@ angular.module('oppia').factory('ResponsesService', [
       }
     };
 
+    var _reduceRuleIndexByOne = function(answerGroupIndex, ruleIndex) {
+      _answerGroups[answerGroupIndex].rules[ruleIndex].inputs.x--;
+    };
+
+    var _makeRuleInvalid = function(answerGroupIndex, ruleIndex) {
+      _answerGroups[answerGroupIndex].rules[ruleIndex].inputs.x = (
+        Number.MAX_SAFE_INTEGER);
```

Comment:
Add comments here to explain rationale.

Also ... I understand your rationale, but I am still a bit uncertain about this. Two questions: (a) is it ever possible to save a rule to the backend when it is in this state, (b) what were we doing/storing previously when this case arose?

Reply:
> (a) is it ever possible to save a rule to the backend when it is in this state

Yes, It is possible to save a rule in this state but we can't publish it.
Could this cause problems?
![anwer_gorup](https://user-images.githubusercontent.com/31371571/76794767-c1874800-67ed-11ea-8aff-7a2b415cc837.png)

> (b) what were we doing/storing previously when this case arose?

Previously, This case was not handled. i.e when choices that corresponded to an answer group were deleted nothing happened.

Reply:
I think we ought to forbid saving to the backend in this state. That way, the exploration is always in a basically-valid/playable state of affairs. Could we add a warning that detects this and asks the user to fix it? If you could add that then I think using the MAX INTEGER thing is fine (I just want to avoid a situation down the road where someone looks at this and wonders what it means, since it's clearly invalid). We should also do a follow-up one-off job to clean up the existing explorations on the server.

Also thanks for the investigation on the previous case, that's good to know.

Reply:
> Could we add a warning that detects this and asks the user to fix it

There is an existing warning for this(see `line 105`, `multiple-choice-input-validation.service.ts`,  which raises a warning when `rules.input.x` becomes greater than the length of the answer choices.
But this warning only prevents the exploration from being published, but still allows to be saved in the backend.        
So is the one-off job still required(because explorations in this state won't be published)?

Reply:
Yes, I think so. I don't see why those should even be saveable... do you see a good reason for that?

Reply:
So for confirming, Should, we still use MAX INTEGER for handling the deletion case.
Also, What exactly should the one-off job look for and accomplish?
Thanks

Reply:
Yes, let's keep MAX_INTEGER.

The exploration migration should iterate through all explorations, find rules that refer to answer choices that no longer exist, and delete those rules. (Check with LaCE team leads for guidance on writing exploration migrations if you're new to this.)

Thanks!